diff --git a/build/moz.configure/rust.configure b/build/moz.configure/rust.configure
index a02557c1a4..9c0296111f 100644
--- a/build/moz.configure/rust.configure
+++ b/build/moz.configure/rust.configure
@@ -163,6 +163,8 @@ def rust_supported_targets(rustc):
                 retry = t[:-len('windows-msvc')] + 'mingw32'
             elif t.endswith('-windows-gnu'):
                 retry = t[:-len('windows-gnu')] + 'mingw32'
+            elif t.endswith('-windows-gnullvm'):
+                retry = t[:-len('windows-gnullvm')] + 'mingw32'
             else:
                 continue
             try:
@@ -217,7 +219,9 @@ def rust_triple_alias(host_or_target, host_or_target_c_compiler):
             # narrow further down using extra information from the build system.
             # - For windows targets, correlate with the C compiler type
             if host_or_target.kernel == 'WINNT':
-                if compiler_info.type in ('gcc', 'clang'):
+                if compiler_info.type == 'clang':
+                    suffix = 'windows-gnullvm'
+                elif compiler_info.type == 'gcc':
                     suffix = 'windows-gnu'
                 else:
                     suffix = 'windows-msvc'
